clippy_lints/src/methods/unused_enumerate_index.rs

```diff
@@ -0,0 +1,90 @@
+use clippy_utils::diagnostics::{multispan_sugg, span_lint, span_lint_and_then};
+use clippy_utils::paths::{CORE_ITER_ENUMERATE_METHOD, CORE_ITER_ENUMERATE_STRUCT};
+use clippy_utils::source::snippet;
+use clippy_utils::{is_trait_method, match_def_path, pat_is_wild};
+use rustc_hir::{Expr, ExprKind, PatKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty::AdtDef;
+use rustc_span::sym;
+
+use crate::loops::UNUSED_ENUMERATE_INDEX;
+
+/// Check for the `UNUSED_ENUMERATE_INDEX` lint outside of loops.
+///
+/// The lint is declared in `clippy_lints/src/loops/mod.rs`. There, the following pattern is
+/// checked:
+/// ```ignore
+/// for (_, x) in some_iter.enumerate() {
+///     // Index is ignored
+/// }
+/// ```
+///
+/// This `check` function checks for chained method calls constructs where we can detect that the
+/// index is unused. Currently, this checks only for the following patterns:
+/// ```ignore
+/// some_iter.enumerate().map_function(|(_, x)| ..)
+/// let x = some_iter.enumerate();
+/// x.map_function(|(_, x)| ..)
+/// ```
+/// where `map_function` is one of `all`, `any`, `filter_map`, `find_map`, `flat_map`, `for_each` or
+/// `map`.
+///
+/// # Preconditons
+/// This function must be called not on the `enumerate` call expression itself, but on any of the
+/// map functions listed above. It will ensure that `recv` is a `std::iter::Enumerate` instance and
+/// that the method call is one of the `std::iter::Iterator` trait.
+///
+/// * `call_expr`: The map function call expression
+/// * `recv`: The receiver of the call
+/// * `closure_arg`: The argument to the map function call containing the closure/function to apply
+pub(super) fn check(cx: &LateContext<'_>, call_expr: &Expr<'_>, recv: &Expr<'_>, closure_arg: &Expr<'_>) {
+    let recv_ty = cx.typeck_results().expr_ty(recv);
+    if let Some(recv_ty_defid) = recv_ty.ty_adt_def().map(AdtDef::did)
+        // If we call a method on a `std::iter::Enumerate` instance
+        && match_def_path(cx, recv_ty_defid, &CORE_ITER_ENUMERATE_STRUCT)
+        // If we are calling a method of the `Iterator` trait
+        && is_trait_method(cx, call_expr, sym::Iterator)
+        // And the map argument is a closure
+        && let ExprKind::Closure(closure) = closure_arg.kind
+        && let closure_body = cx.tcx.hir().body(closure.body)
+        // And that closure has one argument ...
+        && let [closure_param] = closure_body.params
+        // .. which is a tuple of 2 elements
+        && let PatKind::Tuple([index, elem], ..) = closure_param.pat.kind
```

Comment:
If there's an util that can walk through all uses of an expression (either returns an `Iter<Item = &Expr>` or takes a visitor), I could do this in this PR.

Otherwise, this may be a bit out of scope, but I could take care of it in another PR.

Reply:
It should be possible with [`for_each_expr_with_closures`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/visitors/fn.for_each_expr_with_closures.html)

> Otherwise, this may be a bit out of scope, but I could take care of it in another PR.

Yeah, it doesn't need to be done in this PR, just an idea for a potential improvement :)

Reply:
This is done, but proved to be much more verbose than I anticipated. I have created a helper function to check whether the index is used, which is able to return all locations of `x.1` so that we can replace them with `x` when removing `enumerate()`.

This appears to work well, but a lot of the code at that point should be used in the `loops/` part of the lint as well. I don't know if setting my helper's visibility to `pub(crate)` is fine, as it is kind of specific to the `unused_enumerate_index` lint and would "pollute" LSP suggestions, for instance.

I might be wrong, but I think it would be best to merge this PR as-is to get this diff out of the way, and then focus on handling binding patterns.

Reply:
Right, I forgot that that will also require changing all the `x.1`s to `x`. Makes sense to do that separately then :)